testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+
+    Status status = Status.DEADLINE_EXCEEDED.withDescription("It was bound to happen");
+    serverStream.cancel(status);
+    verify(mockServerStreamListener, timeout(1000)).closed(Matchers.same(status));
+    verify(mockClientStreamListener, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    // Presently we can't sent much back to the client in this case. Verify that is the current
+    // behavior for consistency between transports.
+    assertCodeEquals(Status.CANCELLED, statusCaptor.getValue());
+
+    // Second cancellation shouldn't trigger additional callbacks
+    reset(mockServerStreamListener);
+    reset(mockClientStreamListener);
+    serverStream.cancel(status);
+    verify(mockServerStreamListener, never()).closed(any(Status.class));
+    verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
+  }
+
+  @Test
+  public void flowControlPushBack() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
+    assertEquals(methodDescriptor.getFullMethodName(), serverStreamCreation.method);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+    serverStream.writeHeaders(new Metadata());
+
+    Answer<Void> closeStream = new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Exception {
+        Object[] args = invocation.getArguments();
+        ((InputStream) args[0]).close();
+        return null;
+      }
+    };
+
+    String largeMessage;
+    {
+      int size = 1 * 1024;
+      StringBuffer sb = new StringBuffer(size);
+      for (int i = 0; i < size; i++) {
+        sb.append('a');
+      }
+      largeMessage = sb.toString();
+    }
+
+    doAnswer(closeStream).when(mockServerStreamListener).messageRead(any(InputStream.class));
+    serverStream.request(1);
+    verify(mockClientStreamListener, timeout(1000)).onReady();
+    assertTrue(clientStream.isReady());
+    final int maxToSend = 10 * 1024;
+    int clientSent;
+    for (clientSent = 0; clientStream.isReady(); clientSent++) {
+      if (clientSent > maxToSend) {
+        fail("Too many messages sent before isReady() returned false");
+      }
+      clientStream.writeMessage(methodDescriptor.streamRequest(largeMessage));
+      clientStream.flush();
+    }
+    assertTrue(clientSent > 0);
+    for (; clientSent < 5; clientSent++) {
```

Comment:
I'm adding a comment:
> Make sure there are at least a few messages buffered.

If `isReady()` returns `false` almost-immediately (as it will actually do for InProcessTransport), then we want at least a couple messages.